File: lib/chapter3/basic.ts
export type Type =
  | { tag: "Boolean" }
  | { tag: "Number" }
  | { tag: "Func"; params: Param[]; retType: Type };

type Term =
  | { tag: "true" }
  | { tag: "false" }
  | { tag: "if"; cond: Term; thn: Term; els: Term }
  | { tag: "number"; n: number }
  | { tag: "add"; left: Term; right: Term }
  | { tag: "var"; name: string }
  | { tag: "func"; params: Param[]; body: Term }
  | { tag: "call"; func: Term; args: Term[] }
  | { tag: "seq"; body: Term; rest: Term }
  | { tag: "const"; name: string; init: Term; rest: Term };

type Param = { name: string; type: Type };

type TypeEnv = Record<string, Type>;

export function typecheck(t: Term, tyEnv: TypeEnv): Type {
  switch (t.tag) {
    case "true":
      return { tag: "Boolean" };
    case "false":
      return { tag: "Boolean" };
    case "if": {
      const condTy = typecheck(t.cond, tyEnv);
      if (condTy.tag !== "Boolean") throw "boolean expected";
      const thnTy = typecheck(t.thn, tyEnv);
      const elsTy = typecheck(t.els, tyEnv);
      if (thnTy.tag !== elsTy.tag) throw "then and else have different types";
      return thnTy;
    }
    case "number":
      return { tag: "Number" };
    case "add": {
      const leftTy = typecheck(t.left, tyEnv);
      if (leftTy.tag !== "Number") throw "number expected";
      const rightTy = typecheck(t.right, tyEnv);
      if (rightTy.tag !== "Number") throw "number expected";
      return { tag: "Number" };
    }
    case "var": {
      if (tyEnv[t.name] === undefined) throw "variable not found";
      return tyEnv[t.name];
    }
    case "func": {
      const retType = typecheck(t.body, tyEnv);
      return { tag: "Func", params: t.params, retType };
    }
    default: {
      throw "not implemented yet";
      // const _exhaustiveCheck: never = t;
      // throw new Error("unreachable:" + _exhaustiveCheck)
    }
  }
}

export function typeEq(ty1: Type, ty2: Type): boolean {
  switch (ty2.tag) {
    case "Boolean":
      return ty1.tag === "Boolean";
    case "Number":
      return ty1.tag === "Number";
    case "Func": {
      if (ty1.tag !== "Func") return false;
      if (ty1.params.length !== ty2.params.length) return false;
      for (let i = 0; i < ty1.params.length; i++) {
        if (!typeEq(ty1.params[i].type, ty2.params[i].type)) {
          return false;
        }
      }
      if (!typeEq(ty1.retType, ty2.retType)) return false;
      return true;
    }
    default: {
      const _exhaustiveCheck: never = ty2;
      throw new Error("unreachable:" + _exhaustiveCheck)
    }
  }
}

File: lib/chapter3/basic.rb
# rbs_inline: enabled

require_relative "../parser/parser"

# @rbs!
#   type Chapter3::term = { tag: "true" }
#                       | { tag: "false" }
#                       | { tag: "if", cond: Chapter3::term, thn: Chapter3::term, els: Chapter3::term }
#                       | { tag: "number", n: Integer }
#                       | { tag: "add", left: Chapter3::term, right: Chapter3::term }
#                       | { tag: "var", name: string }
#                       | { tag: "func", params: Array[Chapter3::param], body: Chapter3::term }
#                       | { tag: "call", func: Chapter3::term, args: Array[Chapter3::term] }

#   type Chapter3::typ = { tag: "Boolean" }
#                      | { tag: "Number" }
#                      | { tag: "Func", params: Array[Chapter3::param], retType: Chapter3::typ }

# @rbs!
#   type Chapter3::param = { name: string, type: untyped }

#   type Chapter3::param = { name: string, type: Chapter3::typ }

# @rbs!
#   type Chapter3::typeEnv = Hash[string, untyped]

#   type Chapter3::typeEnv = Hash[string, Chapter3::typ]

module Chapter3
  class Checker
    #: (untyped, untyped) -> untyped
    def self.typecheck(t, tyEnv)
      case
      when t[:tag] == "true"
        { tag: "Boolean" }
      when t[:tag] == "false"
        { tag: "Boolean" }
      when t[:tag] == "if"
        condTy = typecheck(t[:cond], tyEnv)
        raise "boolean expected" if condTy[:tag] != "Boolean"
        thnTy = typecheck(t[:thn], tyEnv)
        elsTy = typecheck(t[:els], tyEnv)
        raise "then and else have different types" if thnTy[:tag] != elsTy[:tag]
        return thnTy
      when t[:tag] == "number"
        { tag: "Number" }
      when t[:tag] == "add"
        leftTy = typecheck(t[:left], tyEnv)
        raise "number expected" if leftTy[:tag] != "Number"
        rightTy = typecheck(t[:right], tyEnv)
        raise "number expected" if rightTy[:tag] != "Number"
        { tag: "Number" }
      when t[:tag] == "var"
        raise "variable not found" unless tyEnv.key? t[:name]
        tyEnv[t[:name]]
      when t[:tag] == "func"
        newTyEnv = tyEnv.dup
        t[:params].each do |p|
          newTyEnv[p[:name]] = p[:type] unless p[:type].nil?
        end
        { tag: "Func", params: t[:params], retType: typecheck(t[:body], newTyEnv) }
      when t[:tag] == "call"
        funcTy = typecheck(t[:func], tyEnv)
        raise "function type expected" unless funcTy[:tag] == "Func"
        # funcTy が FuncType であることが自明だが steep は narrowing できないパターン
        raise "wrong number of arguments" if funcTy[:params].size != t[:args].size
        raise "argument type mismatch" if funcTy[:params].zip(t[:args]).any? { |param, argTerm| !typeEq(param[:type], typecheck(argTerm, tyEnv))} # steep:ignore
        retType = funcTy[:retType]
        raise "never raise..." if retType.is_a?(String)
        raise "return type declaration is required" if retType.nil?
        retType
      else
        raise "not implemented"
      end
    end

    #: (untyped, untyped) -> untyped
    def self.typeEq(ty1, ty2)
      case ty2[:tag]
      when "Boolean"
        ty1[:tag] == "Boolean"
      when "Number"
        ty1[:tag] == "Number"
      when "Func"
        return false if ty1[:tag] != "Func"
        return false if ty1[:params].size != ty2[:params].size
        return false if ty1[:params].zip(ty2[:params]).any? { |p1, p2| !typeEq(p1[:type], p2[:type])} # steep:ignore
        return false unless typeEq(ty1[:retType], ty2[:retType]) # steep:ignore
        true
      end
    end
  end
end
